api-server/src/main/java/io/enmasse/api/server/RequestLogger.java

```diff
@@ -4,19 +4,82 @@
  */
 package io.enmasse.api.server;
 
+import io.enmasse.metrics.api.BucketFn;
+import io.enmasse.metrics.api.HistogramMetric;
+import io.enmasse.metrics.api.ScalarMetric;
+import io.enmasse.metrics.api.MetricLabel;
+import io.enmasse.metrics.api.MetricType;
+import io.enmasse.metrics.api.MetricValue;
+import io.enmasse.metrics.api.Metrics;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import javax.ws.rs.container.ContainerRequestContext;
 import javax.ws.rs.container.ContainerRequestFilter;
 import javax.ws.rs.container.ContainerResponseContext;
 import javax.ws.rs.container.ContainerResponseFilter;
-import java.util.concurrent.TimeUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicLongArray;
 
 public class RequestLogger implements ContainerRequestFilter, ContainerResponseFilter {
     private static final Logger log = LoggerFactory.getLogger(RequestLogger.class);
     private static final String PROP_NAME = "request.start";
 
+    // Assume status code above 599 is not used
+    private static final int MAX_CODE = 600;
+    private final AtomicLongArray statusCodeCounter = new AtomicLongArray(MAX_CODE);
+
+    // Buckets:
+    // 0-9 ms
+    // 10-99 ms
+    // 100-999 ms
+    // 1000-9999 ms
+    // 10000-+Inf ms
+    private static final int NUM_BUCKETS = 5;
+    private final AtomicLongArray latencyBucketsCount = new AtomicLongArray(NUM_BUCKETS);
+    private static final Long latencyBucketLimits[] = new Long[] {
```

Comment:
Are we publishing sufficient latency information to Prometheus for it to be useful?  Don't we need to breakdown by HTTP verb and endpoint?

Reply:
Good point, I will add them as labels

Reply:
Actually, I'm a bit worried about the dimentionality here. I think the metrics are useful even without the labels. They indicate that there is an issue, and if one needs to look into the details, they should be in the log.

Reply:
If an app enters a phase where same it causes say lots of more creation events, response latencies will rise as PUTs normaly take longer than GETs .  This would be normal.  The human operator looking a Prometheus won't have the information to percieve that and will probably just flag the spike in response times as abnormal.

Is dimentionality actually an issue?  I think we might be talking about no more than 3  endpoints (address-space, address, user) * 4 verbs.  This seems to fall within Prometheus best practice and within the cardinality warnings within the docs.

I'm being influenced by what I see here:
https://github.com/kubernetes/apiserver/blob/master/pkg/endpoints/metrics/metrics.go#L93

Reply:
OTOH as we know EnMasse api-server's days are numbered, maybe I should worry less :)